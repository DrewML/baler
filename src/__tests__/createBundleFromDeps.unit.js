const getModuleUnderTest = () => {
    jest.resetModules();
    return require('../createBundleFromDeps');
};
const stubDateISOString = () => {
    // We include a timestamp in the generated bundles. Force it to be
    // consistent so snapshot tests work as expected
    const KNOWN_STAMP =
        'Mon Sep 16 2019 17:46:16 GMT-0500 (Central Daylight Time)';
    jest.spyOn(global.Date.prototype, 'toISOString').mockImplementation(
        jest.fn().mockReturnValue(KNOWN_STAMP),
    );
};
const stubFilePaths = mappings => {
    jest.mock('../fsPromises', () => ({
        readFile: async path => {
            if (mappings[path]) {
                return mappings[path];
            }

            throw new Error(`Failed reading ${path} in mock fs`);
        },
    }));
};

test('html file loaded with text plugin gets wrapped as text module', async () => {
    stubFilePaths({
        'root/template.html': '<div>Hello World</div>',
    });
    stubDateISOString();
    const { createBundleFromDeps } = getModuleUnderTest();

    const results = await createBundleFromDeps(
        'core-bundle',
        ['text!template.html'],
        'root',
        {},
        'Vendor/theme',
    );

    expect(results.bundle).toMatchSnapshot();
});

test('non-html file loaded with text plugin gets wrapped as text module', async () => {
    stubFilePaths({
        'root/js-translation.json': '{"hello": "world" }',
    });
    stubDateISOString();
    const { createBundleFromDeps } = getModuleUnderTest();

    const results = await createBundleFromDeps(
        'core-bundle',
        ['text!js-translation.json'],
        'root',
        {},
        'Vendor/theme',
    );

    expect(results.bundle).toMatchSnapshot();
});

test('named AMD module keeps its name', async () => {
    stubFilePaths({
        'root/main.js': `
            define('main', function() {
                console.log('hello world');
            });
        `,
    });
    stubDateISOString();
    const { createBundleFromDeps } = getModuleUnderTest();

    const results = await createBundleFromDeps(
        'core-bundle',
        ['main'],
        'root',
        {},
        'Vendor/theme',
    );

    expect(results.bundle).toMatchSnapshot();
});

test('anonymous AMD module gets a name', async () => {
    stubFilePaths({
        'root/main.js': `
            define(function() {
                console.log('hello world');
            });
        `,
    });
    stubDateISOString();
    const { createBundleFromDeps } = getModuleUnderTest();

    const results = await createBundleFromDeps(
        'core-bundle',
        ['main'],
        'root',
        {},
        'Vendor/theme',
    );

    expect(results.bundle).toMatchSnapshot();
});

test('non-AMD module with shim gets converted to AMD module', async () => {
    stubFilePaths({
        'root/main.js': `
            $(function() {
                // hey look I'm using jQuery outside an AMD wrapper!
            });
        `,
    });
    stubDateISOString();
    const { createBundleFromDeps } = getModuleUnderTest();

    const results = await createBundleFromDeps(
        'core-bundle',
        ['main'],
        'root',
        {
            shim: {
                main: ['jquery'],
            },
        },
        'Vendor/theme',
    );

    expect(results.bundle).toMatchSnapshot();
});

test('non-AMD module without shim gets pseudo amd-wrapper for compat', async () => {
    stubFilePaths({
        'root/main.js': `
            $(function() {
                // hey look I'm using jQuery outside an AMD wrapper!
            });
        `,
    });
    stubDateISOString();
    const { createBundleFromDeps } = getModuleUnderTest();

    const results = await createBundleFromDeps(
        'core-bundle',
        ['main'],
        'root',
        {},
        'Vendor/theme',
    );

    expect(results.bundle).toMatchSnapshot();
});
