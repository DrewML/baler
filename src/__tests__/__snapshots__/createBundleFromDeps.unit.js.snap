// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anonymous AMD module gets a name 1`] = `
"/* Generated by @magento/baler - Mon Sep 16 2019 17:46:16 GMT-0500 (Central Daylight Time) */


            define('main', function() {
                console.log('hello world');
            });
        "
`;

exports[`html file loaded with text plugin gets wrapped as text module 1`] = `
"/* Generated by @magento/baler - Mon Sep 16 2019 17:46:16 GMT-0500 (Central Daylight Time) */

define('text!template.html', function() {
    return '<div>Hello World</div>';
});"
`;

exports[`named AMD module keeps its name 1`] = `
"/* Generated by @magento/baler - Mon Sep 16 2019 17:46:16 GMT-0500 (Central Daylight Time) */


            define('main', function() {
                console.log('hello world');
            });
        "
`;

exports[`non-AMD module with shim gets converted to AMD module 1`] = `
"/* Generated by @magento/baler - Mon Sep 16 2019 17:46:16 GMT-0500 (Central Daylight Time) */

define('main', [\\"jquery\\"], function() {
        // Shimmed by @magento/baler
        (function() {
            
            $(function() {
                // hey look I'm using jQuery outside an AMD wrapper!
            });
        ;
        })();
        return window['undefined'];
    });"
`;

exports[`non-AMD module without shim gets pseudo amd-wrapper for compat 1`] = `
"/* Generated by @magento/baler - Mon Sep 16 2019 17:46:16 GMT-0500 (Central Daylight Time) */

define('main', function() {
    // baler-injected stub for non-AMD module (no shim config was found for this module)
});
// Original code for non-AMD module main

            $(function() {
                // hey look I'm using jQuery outside an AMD wrapper!
            });
        "
`;

exports[`non-html file loaded with text plugin gets wrapped as text module 1`] = `
"/* Generated by @magento/baler - Mon Sep 16 2019 17:46:16 GMT-0500 (Central Daylight Time) */

define('text!js-translation.json', function() {
    return '{\\"hello\\": \\"world\\" }';
});"
`;
